### PR TITLE
Add chunk encoding for float histogram

### PR DESCRIPTION
**Pull request by codesome**
This PR adds a new chunk encoding for the float histogram. This is a prerequisite to test out https://github.com/prometheus/prometheus/pull/11522/ (it includes some work from that PR for the chunk). @marctc and I decided it is best to have it as a separate PR for easy review.

### CODE REVIEW

**Comment by beorn7** [comment]
On my review queue…

File: tsdb/chunkenc/float_histogram.go [hunk 1]
```diff
@@ -0,0 +1,679 @@
...
+
+		it.numRead++
+		return ValFloatHistogram
+	}
+
+	// The case for the 2nd sample with single deltas is implicitly handled correctly with the double delta code,
+	// so we don't need a separate single delta logic for the 2nd sample.
+
+	// Recycle bucket slices that have not been returned yet. Otherwise,
+	// copy them.
```
Comment (beorn7) [inline tsdb/chunkenc/float_histogram.go:612]: Perhaps mention here that we can always recycle the slices for leading and trailing bits as they are never returned to the caller.

File: tsdb/chunkenc/histogram_meta.go [hunk 2]
```diff
@@ -280,19 +280,23 @@ loop:
...
+		interj int // The next interjection to process.
 	)
 	for i, d := range in {
 		if interj < len(interjections) && i == interjections[interj].pos {
 
 			// We have an interjection!
 			// Add interjection.num new delta values such that their
 			// bucket values equate 0.
-			out[j] = int64(-v)
+			if deltas {
```
Comment (beorn7) [inline tsdb/chunkenc/histogram_meta.go:295]: The code comment above needs an update for the delta==false case.

**Review by beorn7 — CHANGES_REQUESTED** [review]
Looks great in general. But found some things to comment upon. :)

**Comment by beorn7** [comment]
To reiterate: I think we should handle the gauge histogram case later (see detailed comments above). Let's first get the float histogram (for counter-like float histograms) done, and then think about the gauge histogram case (without counter resets, with disappearing buckets, and such), as this needs to be solved for integer histograms, too.

Another thing we can think about later: If a `FloatHistogram` is effectively an integer histogram (because all the floats in it can be represented by ints), we can store it more efficiently as a normal (integer) `Histogram`. However, this adds the complication that we need to re-encode or at least start a new chunk if suddenly the next histogram for the same series is an "actual" `FloatHistogram`. This optimization could be done behind the scenes, or we can "leave it to the caller", i.e. implement it in the ingestion part. This deserves some thought, as we should probably do the same thing for conventional float samples. Many of those are effectively integers and could be stored much more efficiently. That's true for something like request counters but even more so for the many many time series that are always 1 or always 0, or those that are either 1 or 0. (For those, we could introduce even more efficient RLE like encodings, see #10275 .) All of this is out of scope of this PR, of course. I just mention it to justify why we don't want to do this kind of optimization right here.

File: tsdb/chunkenc/float_histogram.go [hunk 3]
```diff
@@ -0,0 +1,679 @@
...
+	}
+
+	if h.Schema != a.schema || h.ZeroThreshold != a.zThreshold {
+		return
+	}
+
+	var ok bool
+	positiveInterjections, ok = compareSpans(a.pSpans, h.PositiveSpans)
+	if !ok {
+		bucketsMissing = true
```
[the remark this answers is not in the document]
Comment (codesome) [reply tsdb/chunkenc/float_histogram.go:254]: Bucket missing is as good as a counter reset, so I guess you mean only a name change here while adding counter reset logic everywhere.

File: tsdb/chunkenc/float_histogram.go [hunk 4]
```diff
@@ -0,0 +1,679 @@
...
+		return
+	}
+	negativeInterjections, ok = compareSpans(a.nSpans, h.NegativeSpans)
+	if !ok {
+		bucketsMissing = true
+		return
+	}
+
```
[the remark this answers is not in the document]
Comment (codesome) [reply tsdb/chunkenc/float_histogram.go:262]: For it to be efficient, it cannot be generics :/
The "oldBuckets" slice for the FloatHistogram is []xorValue and not []float64, and creating a new slice for this check everytime will be inefficient.

File: tsdb/chunkenc/float_histogram.go [hunk 5]
```diff
@@ -0,0 +1,679 @@
...
+	it.br = newBReader(b[3:])
+	it.numTotal = binary.BigEndian.Uint16(b)
+	it.numRead = 0
+
+	it.t, it.tDelta = 0, 0
+	it.cnt, it.zCnt, it.sum = xorValue{}, xorValue{}, xorValue{}
+
+	it.atFloatHistogramCalled = false
+	it.pBuckets, it.nBuckets = nil, nil
+	it.pBucketsLeading, it.pBucketsTrailing, it.nBucketsLeading, it.nBucketsTrailing = nil, nil, nil, nil
```
[the remark this answers is not in the document]
Comment (codesome) [reply tsdb/chunkenc/float_histogram.go:513]: I don't know in depth how these are used, so I will take your word for it.

File: tsdb/chunkenc/float_histogram.go [hunk 6]
```diff
@@ -0,0 +1,679 @@
...
+		}
+
+		numPBuckets, numNBuckets := countSpans(h.PositiveSpans), countSpans(h.NegativeSpans)
+		if numPBuckets > 0 {
+			a.pBuckets = make([]xorValue, numPBuckets)
+			for i := 0; i < numPBuckets; i++ {
+				a.pBuckets[i] = xorValue{
+					value:    h.PositiveBuckets[i],
+					leading:  0xff,
+					trailing: 0xff,
```
[the remark this answers is not in the document]
Comment (codesome) [reply tsdb/chunkenc/float_histogram.go:308]: No idea. Will remove it.

File: tsdb/chunkenc/float_histogram.go [hunk 7]
```diff
@@ -0,0 +1,679 @@
...
+		} else {
+			a.pBuckets = nil
+		}
+		if numNBuckets > 0 {
+			a.nBuckets = make([]xorValue, numNBuckets)
+			for i := 0; i < numNBuckets; i++ {
+				a.nBuckets[i] = xorValue{
+					value:    h.NegativeBuckets[i],
+					leading:  0xff,
+					trailing: 0xff,
```
[the remark this answers is not in the document]
Comment (codesome) [reply tsdb/chunkenc/float_histogram.go:320]: No idea. Will remove it.

File: tsdb/chunkenc/float_histogram_test.go [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/prometheus/prometheus/model/histogram"
+)
+
+func TestFloatHistogramChunkSameBuckets(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (codesome) [reply tsdb/chunkenc/float_histogram_test.go:24]: Good point! I switched to the improved tests and already found bugs

File: tsdb/chunkenc/float_histogram.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (beorn7) [reply tsdb/chunkenc/float_histogram.go:279]: Too bad. But thanks for catching.

**Review by beorn7 — APPROVED** [review]
Looks correct and complete to me now. @marctc if you have anything to add, please let us know.

**Pull request merged**